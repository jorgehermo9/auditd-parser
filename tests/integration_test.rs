use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
