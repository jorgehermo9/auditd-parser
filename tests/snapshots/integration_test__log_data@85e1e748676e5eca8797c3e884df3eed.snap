---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/linux-audit-parser-rs/line-syscall.log:1"
expression: result
info: "type=SYSCALL msg=audit(1615114232.375:15558): arch=c000003e syscall=59 success=yes exit=0 a0=63b29337fd18 a1=63b293387d58 a2=63b293375640 a3=fffffffffffff000 items=2 ppid=10883 pid=10884 auid=1000 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=pts1 ses=1 comm=\"whoami\" exe=\"/usr/bin/whoami\" key=(null)\u001dARCH=x86_64 SYSCALL=execve AUID=\"user\" UID=\"root\" GID=\"root\" EUID=\"root\" SUID=\"root\" FSUID=\"root\" EGID=\"root\" SGID=\"root\" FSGID=\"root\""
---
{
  "record_type": "SYSCALL",
  "timestamp": 1615114232375,
  "id": 15558,
  "fields": {
    "a0": "63b29337fd18",
    "a1": "63b293387d58",
    "a2": "63b293375640",
    "a3": "fffffffffffff000",
    "arch": "c000003e",
    "auid": 1000,
    "comm": "whoami",
    "egid": "root",
    "euid": "root",
    "exe": "/usr/bin/whoami",
    "exit": 0,
    "fsgid": "root",
    "fsuid": "root",
    "gid": "root",
    "items": "2",
    "key": "(null)",
    "pid": "10884",
    "ppid": "10883",
    "ses": "1",
    "sgid": "root",
    "success": true,
    "suid": "root",
    "syscall": "59",
    "tty": "pts1",
    "uid": "root"
  },
  "enrichment": {
    "ARCH": "x86_64",
    "AUID": "user",
    "EGID": "root",
    "EUID": "root",
    "FSGID": "root",
    "FSUID": "root",
    "GID": "root",
    "SGID": "root",
    "SUID": "root",
    "SYSCALL": "execve",
    "UID": "root"
  }
}
